src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java

```diff
@@ -1010,12 +1004,24 @@ NameAssignedState assignName(final SecurityIdentity capturedIdentity, final Mech
     NameAssignedState assignName(final SecurityIdentity capturedIdentity, final MechanismConfiguration mechanismConfiguration, final MechanismRealmConfiguration mechanismRealmConfiguration, Principal originalPrincipal, final Evidence evidence, final IdentityCredentials privateCredentials, final IdentityCredentials publicCredentials, final boolean exclusive) throws RealmUnavailableException {
         final SecurityDomain domain = capturedIdentity.getSecurityDomain();
         final Principal preRealmPrincipal = rewriteAll(originalPrincipal, mechanismRealmConfiguration.getPreRealmRewriter(), mechanismConfiguration.getPreRealmRewriter(), domain.getPreRealmRewriter());
+        if (preRealmPrincipal == null) {
+            log.tracef("Unable to rewrite principal [%s] by pre-realm rewritters", originalPrincipal);
+            return new NameAssignedState(capturedIdentity, null, RealmIdentity.NON_EXISTENT, preRealmPrincipal, mechanismConfiguration, mechanismRealmConfiguration, privateCredentials, publicCredentials);
```

Comment:
If the rewrite fails at this stage, it means the user name is invalid and there can be no assigned name.  So the resultant stage should be a subclass of UnassignedState - possibly an InvalidNameState even, if we need special behavior in this case (for example we would need authorization to fail).